File: 강의 자료/02-알고리즘/12-트리/문제별 코드/20364-부동산 다툼/solution.py
import sys
sys.setrecursionlimit(500005)
si = sys.stdin.readline

n, Q = map(int, si().split())


def pro():
    # 사고 싶은 땅에서 시작해서 루트 땅까지 올라오면서 정답을 찾으면 편하다.
    # 왜냐하면 부모 노드는 찾기 쉽기 때문이다.
    estate = [False] * (n + 1)
    for _ in range(Q):
        x = int(si())
        y = x
        res = 0
        while x:
            if estate[x]: res = x
            x //= 2
        estate[y] = True
        print(res)


if __name__ == "__main__":
    pro()

File: 강의 자료/02-알고리즘/12-트리/문제별 코드/20364-부동산 다툼/solution.java
import java.io.*;
import java.lang.reflect.Array;
import java.util.*;

public class Main {
    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int n, Q;
    static boolean[] estate;

    static void input() throws IOException {
        n = scan.nextInt();
        Q = scan.nextInt();
        estate = new boolean[n + 1];
    }

    static void pro() {
        // 사고 싶은 땅에서 시작해서 루트 땅까지 올라오면서 정답을 찾으면 편하다.
        // 왜냐하면 부모 노드는 찾기 쉽기 때문이다.
        while (Q-- > 0){
            int x = scan.nextInt(), res = 0;
            int y = x;
            while (x > 0) {
                if (estate[x]) res = x;
                x >>= 1;
            }
            estate[y] = true;
            sb.append(res).append('\n');
        }
        System.out.println(sb);
    }

    public static void main(String[] args) throws IOException {
        input();
        pro();
    }


    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}
